port/linux/omrvmem.c

```diff
@@ -1228,6 +1228,218 @@ reserve_memory_with_mmap(struct OMRPortLibrary *portLibrary, void *address, uint
+ *  @param void                         *regionAddresses[]  [in] Addresses to be double mapped
+ *  @param uintptr_t                    regionsCount        [in] Number of regions to be double mapped
+ *  @param uintptr_t                    regionSize          [in] Size of each region
+ *  @param uintptr_t                    byteAmount          [in] Total size to allocate for contiguous block of memory
+ *  @param struct J9PortVmemIdentifier  *oldIdentifier      [in] Old Identifier containing file descriptor
+ *  @param struct J9PortVmemIdentifier  *newIdentifier      [out] New Identifier for new block of memory. The structure to be updated
+ *  @param uintptr_t                    mode                [in] Access mode
+ *  @param uintptr_t                    pageSize            [in] Constant describing page size
+ *  @param OMRMemCategory               *category           [in] Memory allocation category
+ *  @param void                         *preferredAddress    [in] Prefered address of contiguous region to be double mapped
+ * 
+ * @return pointer to contiguous region to which regions were double mapped into, NULL is returned if unsuccessful
+ */
+
+void *
+omrvmem_create_double_mapped_region(struct OMRPortLibrary *portLibrary, void* regionAddresses[], uintptr_t regionsCount, uintptr_t regionSize, uintptr_t byteAmount, struct J9PortVmemIdentifier *oldIdentifier, struct J9PortVmemIdentifier *newIdentifier, uintptr_t mode, uintptr_t pageSize, OMRMemCategory *category, void *preferredAddress)
+{
+	Trc_PRT_double_map_regions_Create_Entry((void *)regionAddresses, regionsCount, regionSize, byteAmount, pageSize, preferredAddress);
+	int protectionFlags = PROT_READ | PROT_WRITE;
+	int flags = MAP_PRIVATE | MAP_ANON;
+	int fd = -1;
+	void* contiguousMap = NULL;
+	BOOLEAN successfulContiguousMap = FALSE;
+	BOOLEAN shouldUnmapAddr = FALSE;
+	/* All regions are fully double mapped. Partial double mapping is not supported yet. */
+	Assert_PRT_true((regionsCount * regionSize) == byteAmount);
+
+	if (0 != (OMRPORT_VMEM_MEMORY_MODE_COMMIT & mode)) {
+		protectionFlags = get_protectionBits(mode);
+	}
+	if (0 != (OMRPORT_VMEM_MEMORY_MODE_MMAP_HUGE_PAGES & oldIdentifier->mode)) {
+		flags |= MAP_HUGETLB;
+	}
+
+	/* Create contiguous region if we don't have one already */
+	if (NULL == preferredAddress) {
+		contiguousMap = mmap(
+				NULL,
+				byteAmount,
+				protectionFlags,
+				flags,
+				fd,
+				0);
+		Trc_PRT_double_map_regions_Create_GenerateContiguousAddress(contiguousMap);
+	} else {
+		contiguousMap = preferredAddress;
+		Trc_PRT_double_map_regions_Create_UsingPreferredAddress(preferredAddress);
+	}
+
+	if ((MAP_FAILED == contiguousMap) || (NULL == contiguousMap)) {
+		Trc_PRT_double_map_regions_Create_Failure();
+		contiguousMap = NULL;
+		portLibrary->error_set_last_error_with_message(portLibrary, OMRPORT_ERROR_VMEM_OPFAILED, "Failed to map contiguous block of memory");
```

Comment:
Seems it could be useful to retain the OS errno. How about creating new portable error codes instead of using OPFAILED for every failure.